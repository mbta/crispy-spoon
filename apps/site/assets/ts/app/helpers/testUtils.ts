/* eslint-disable */
import toJson, { Json } from "enzyme-to-json";
import { ReactWrapper } from "enzyme";
import {
  IndexedRouteStop,
  StopTree
} from "../../schedule/components/__schedule";

export const testRouteStopListFromStopTree = (
  tree: StopTree
): IndexedRouteStop[] =>
  Object.values(tree.byId).map((node, index) => ({
    ...node.value,
    routeIndex: index
  }));

export const createReactRoot = (): void => {
  document.body.innerHTML =
    '<div><div id="react-root"><div id="test"></div></div></div>';
};

export const enzymeToJsonWithoutProps = (wrapper: ReactWrapper<any>): Json =>
  toJson(wrapper, {
    noKey: false,
    map: info => {
      if (!info.props) {
        return info;
      }

      // @ts-ignore is a <style></style> declaration
      if (info.node.type === "style") {
        return info;
      }
      // @ts-ignore is an element in a <noscript></noscript> declaration
      if (info.node.nodeType == undefined) {
        // @ts-ignore
        return info;
      }
      //@ts-ignore is a React component (not HTML element)
      if (info.node.nodeType.toString().includes("function")) {
        // @ts-ignore
        info.props = undefined;
      }

      return info;
    }
  });

export const toJsonEnzyme = (wrapper: ReactWrapper<any>): Json =>
  toJson(wrapper, {
    noKey: false
  });
