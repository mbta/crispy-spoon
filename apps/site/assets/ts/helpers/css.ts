import {
  isABusRoute,
  isACommuterRailRoute,
  isAGreenLineRoute,
  isASilverLineRoute
} from "../models/route";
import { Route } from "../__v3api";

export const routeToCSSClass = (route: string): string =>
  route.toLowerCase().replace(/[ _]/g, "-");

export const busClass = (route: Route): string =>
  isABusRoute(route) && !isASilverLineRoute(route) ? "bus-route-sign" : "";

export const routeToModeName = (route: Route): string => {
  if (isACommuterRailRoute(route)) return "commuter-rail";
  if (route.type === 4) return "ferry";
  if (route.id === "Red" || route.id === "Mattapan") return "red-line";
  if (route.id === "Orange") return "orange-line";
  if (route.id === "Blue") return "blue-line";
  if (isAGreenLineRoute(route)) return "green-line";
  if (isASilverLineRoute(route)) return "silver-line";
  if (isABusRoute(route)) return "bus";
  return "unknown";
};

export const routeBgClass = (route: Route): string =>
  `u-bg--${routeToModeName(route)}`;
