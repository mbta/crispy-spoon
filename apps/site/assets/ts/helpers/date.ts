// Returns a date in the form of "August 7, 2019"
export const formattedDate = (unformatted: string): string => {
  const [year, month, day] = unformatted
    .split(/-/)
    .map(part => Number.parseInt(part, 10));

  // Remember that months in JS are 0-indexed for some reason, hence "month - 1".
  const parsedDate = new Date(year, month - 1, day);
  return parsedDate.toLocaleDateString("en-US", {
    year: "numeric",
    month: "long",
    day: "numeric",
    timeZone: "UTC"
  });
};

// Returns a date in the form of "Aug 7"
export const shortDate = (date: Date): string =>
  date.toLocaleDateString("en-US", {
    month: "short",
    day: "numeric",
    timeZone: "UTC"
  });

export default formattedDate;
