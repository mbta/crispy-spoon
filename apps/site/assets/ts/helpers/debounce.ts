// Debounce is meant to be used with event listeners. Use it to delay a call
// until after a burst of events, such as keyboard input.
//
//   usage:
//
//     const callback = (ev: Event) => {}
//     document.addEventListener("keyup", debounce(callback, 250))
//

const debounce = (callback: Function, delay: number): EventListener => {
  let timer: number | undefined;

  return (...args) => {
    window.clearTimeout(timer);
    timer = window.setTimeout(() => callback(...args), delay);
  };
};

export const debouncePromise = (
  callback: Function,
  delay: number
): Function => {
  let timerId: number | undefined;

  return function debounced(...args: unknown[]) {
    if (timerId) {
      window.clearTimeout(timerId);
    }

    return new Promise(resolve => {
      timerId = window.setTimeout(() => resolve(callback(...args)), delay);
    });
  };
};

export default debounce;
