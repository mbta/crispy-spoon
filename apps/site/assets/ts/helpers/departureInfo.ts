import { concat, difference, keyBy, keys, map, sortBy } from "lodash";
import { PredictionWithTimestamp } from "../models/perdictions";
import { ScheduleWithTimestamp } from "../models/schedules";
import { isCancelled, isDelayed } from "./prediction-helpers";
import { DepartureInfo } from "../models/departureInfo";

const departureInfoToTime = (departureInfo: DepartureInfo): Date => {
  // If there isn't a prediction there should be a schedule
  return departureInfo.prediction
    ? departureInfo.prediction.time
    : departureInfo.schedule!.time;
};

const mergeIntoDepartureInfo = (
  schedules: ScheduleWithTimestamp[],
  predictions: PredictionWithTimestamp[]
): DepartureInfo[] => {
  const schedulesByTripId = keyBy(schedules, sch => sch.trip.id);
  const predictionsByTripId = keyBy(predictions, prd => prd.trip.id);

  const scheduleTripIds = keys(schedulesByTripId);
  const predictionTripIds = keys(predictionsByTripId);

  // All the schedules with and without predictions
  const scheduleDirectionInfo = map(
    schedulesByTripId,
    (schedule, tripId): DepartureInfo => {
      // prediction could be undefined
      const prediction = predictionsByTripId[tripId];
      return {
        prediction: prediction,
        schedule: schedule,
        isCancelled: isCancelled(prediction),
        isDelayed: isDelayed(prediction, schedule)
      };
    }
  );

  const unmatchedPredictionTripIds = difference(
    predictionTripIds,
    scheduleTripIds
  );
  // All the predictions with no schedules
  const predictionDirectionInfo = map(
    unmatchedPredictionTripIds,
    (tripId: string): DepartureInfo => {
      const prediction = predictionsByTripId[tripId];
      return {
        prediction: prediction,
        isCancelled: isCancelled(prediction)
      } as DepartureInfo;
    }
  );

  const departureInfos = concat(scheduleDirectionInfo, predictionDirectionInfo);
  const sortedDepartureInfo = sortBy(departureInfos, di => {
    // prioritize sorting by predictions
    if (di.prediction) {
      return di.prediction.time;
    }
    return di.schedule?.time;
  });

  return sortedDepartureInfo;
};

export { mergeIntoDepartureInfo, departureInfoToTime };
