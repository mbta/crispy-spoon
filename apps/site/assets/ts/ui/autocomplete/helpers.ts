import { omit } from "lodash";
import { OnSubmitParams } from "@algolia/autocomplete-core";
import { OnStateChangeProps } from "@algolia/autocomplete-js";
import {
  ContentItem,
  Item,
  LocationItem,
  PopularItem,
  RouteItem,
  StopItem
} from "./__autocomplete";
import { isLGDown } from "../../helpers/media-breakpoints";

export function isStopItem(x: Item): x is StopItem {
  return Object.keys(x).includes("stop");
}

export function isRouteItem(x: Item): x is RouteItem {
  return Object.keys(x).includes("route");
}

export function isContentItem(x: Item): x is ContentItem {
  return Object.keys(x).includes("_content_type");
}

export const getTitleAttribute = (item: Item): string[] => {
  if (isStopItem(item)) {
    return ["stop", "name"];
  }
  if (isRouteItem(item)) {
    return ["route", "name"];
  }

  // @ts-ignore
  // eslint-disable-next-line no-underscore-dangle
  return Object.keys(item._highlightResult);
};

const navStateChange: (props: OnStateChangeProps<Item>) => void = ({
  state
}) => {
  // grey out the page and disable scrolling when search is open
  if (isLGDown()) {
    if (state.isOpen) {
      document.documentElement.dataset.navOpen = "true";
    } else {
      delete document.documentElement.dataset.navOpen;
    }
  }
};

export const STATE_CHANGE_HANDLERS: Record<
  string,
  (props: OnStateChangeProps<Item>) => void
> = {
  nav: navStateChange
};

export const SUBMIT_HANDLERS: Record<
  string,
  (props: OnSubmitParams<Item>) => void
> = {
  to_search_page({ state }) {
    window.Turbolinks.visit(`/search?query=${state.query}`);
  }
};

export type WithUrls<T> = T & { urls: Record<string, string> };

// The backend returns all possible URLs, use urlType to get the desired one
export const itemWithUrl = (
  initialItem: WithUrls<LocationItem | PopularItem>,
  urlType: string
): Omit<typeof initialItem, "urls"> => {
  const item = omit(initialItem, "urls");
  item.url = initialItem.urls[urlType];
  return item;
};
