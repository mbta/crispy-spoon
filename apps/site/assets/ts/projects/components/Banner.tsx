import React, { ReactElement } from "react";
import { Teaser as Project, Route } from "../../__cms";
import { formattedDate } from "../../helpers/date";
import { routeToCSSClass } from "../../helpers/css";

interface Props {
  banner: Project | null;
  placeholderImageUrl: string;
}

export const cmsRouteToClass = (route: Route): string => {
  if (route.id === "Green") return "green-line";
  if (route.id === "silver_line") return "silver-line";
  if (route.id === "mattapan") return "mattapan";
  if (route.mode === "commuter_rail") return "commuter-rail";
  if (route.group === "custom" && route.mode === null) return "unknown";
  if (route.group === "custom") return routeToCSSClass(route.mode!);
  if (route.group === "mode" && route.id !== null)
    return routeToCSSClass(route.id);
  if (route.group === "line") return `${routeToCSSClass(route.id)}-line`;
  return "unknown";
};

const bannerRouteClass = ({ routes }: Project): string =>
  cmsRouteToClass(routes[0]);

const bannerBgClass = (banner: Project): string => {
  if (banner.routes.length === 0) return "u-bg--unknown";
  return bannerRouteClass(banner);
};

const bannerContentClass = (banner: Project): string =>
  `c-banner__content c-banner__content--responsive-side-by-side c-banner__content--left u-bg--${bannerBgClass(
    banner
  )}`;

const bannerUpdated = (banner: Project): ReactElement<HTMLElement> => (
  <span className="c-banner__date">Updated {formattedDate(banner.date)}</span>
);

const BannerContent = ({
  banner
}: {
  banner: Project;
}): ReactElement<HTMLElement> => (
  <div className={bannerContentClass(banner)}>
    <div className="c-banner__top">
      <h2 className="h2 c-banner__title c-banner__title--default">
        {banner.title}
      </h2>
    </div>
    <div className="c-banner__bottom">{bannerUpdated(banner)}</div>
  </div>
);

const bannerImageURL = (banner: Project, placeholderImageUrl: string): string =>
  banner.image ? banner.image.url : placeholderImageUrl;

const bannerImageAlt = (banner: Project): string =>
  banner.image ? banner.image.alt : "MBTA logo";

const BannerXS = ({
  banner,
  placeholderImageUrl
}: {
  banner: Project;
  placeholderImageUrl: string;
}): ReactElement<HTMLElement> => (
  <div className="hidden-sm-up">
    <div className="row">
      <div
        className="c-banner__image c-banner__image--responsive c-banner__image--default"
        style={{
          backgroundImage: `url(${bannerImageURL(banner, placeholderImageUrl)})`
        }}
      >
        <div className="sr-only">{bannerImageAlt(banner)}</div>
      </div>
      <BannerContent banner={banner} />
    </div>
  </div>
);

const BannerSideBySide = ({
  banner,
  placeholderImageUrl
}: {
  banner: Project;
  placeholderImageUrl: string;
}): ReactElement<HTMLElement> => (
  <div className="c-banner__image c-banner__image--responsive-side-by-side c-banner--responsive-no-margin c-banner__image--default">
    <BannerContent banner={banner} />
    <img
      className="c-banner__image--by-side"
      src={bannerImageURL(banner, placeholderImageUrl)}
      alt={bannerImageAlt(banner)}
    />
    <div className="sr-only">{bannerImageAlt(banner)}</div>
  </div>
);

const BannerSMUp = ({
  banner,
  placeholderImageUrl
}: {
  banner: Project;
  placeholderImageUrl: string;
}): ReactElement<HTMLElement> => (
  <div className="hidden-xs-down hidden-lg-up">
    <div className="row">
      <BannerSideBySide
        banner={banner}
        placeholderImageUrl={placeholderImageUrl}
      />
    </div>
  </div>
);

const BannerLGUp = ({
  banner,
  placeholderImageUrl
}: {
  banner: Project;
  placeholderImageUrl: string;
}): ReactElement<HTMLElement> => (
  <div className="hidden-md-down">
    <BannerSideBySide
      banner={banner}
      placeholderImageUrl={placeholderImageUrl}
    />
  </div>
);

const Banner = ({
  banner,
  placeholderImageUrl
}: Props): ReactElement<HTMLElement> | null => {
  if (!banner) {
    return null;
  }

  return (
    <>
      <h2 className="c-projects-header__subheader">Featured Projects</h2>
      <a
        href={banner.path}
        className="c-banner c-banner--responsive c-banner--lg-9 c-banner--no-margin-top c-banner--default"
      >
        <BannerXS banner={banner} placeholderImageUrl={placeholderImageUrl} />
        <BannerSMUp banner={banner} placeholderImageUrl={placeholderImageUrl} />
        <BannerLGUp banner={banner} placeholderImageUrl={placeholderImageUrl} />
      </a>
    </>
  );
};

export default Banner;
