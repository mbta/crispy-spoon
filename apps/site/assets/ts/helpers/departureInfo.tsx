import React from "react";
import {
  chain,
  concat,
  difference,
  find,
  keyBy,
  keys,
  map,
  sortBy
} from "lodash";
import { PredictionWithTimestamp } from "../models/predictions";
import { ScheduleWithTimestamp } from "../models/schedules";
import { isCancelled, isDelayed, isSkipped } from "./prediction-helpers";
import { DepartureInfo } from "../models/departureInfo";
import {
  isABusRoute,
  isACommuterRailRoute,
  isSubwayRoute
} from "../models/route";
import { DirectionId, Route } from "../__v3api";
import DisplayTime from "../stop/components/DisplayTime";
import { getInfoKey } from "../stop/models/displayTimeConfig";
import { RoutePatternWithPolyline } from "../models/route-patterns";

export const SUBWAY = "subway";
export const BUS = "bus";
export const COMMUTER_RAIL = "commuter_rail";
export const FERRY = "ferry";

const toRouteMode = (
  route: Route
): typeof SUBWAY | typeof BUS | typeof COMMUTER_RAIL | typeof FERRY => {
  if (isSubwayRoute(route)) {
    return SUBWAY;
  }
  if (isABusRoute(route)) {
    return BUS;
  }
  if (isACommuterRailRoute(route)) {
    return COMMUTER_RAIL;
  }
  return FERRY;
};

const departureInfoToTime = (departureInfo: DepartureInfo): Date => {
  // If there isn't a prediction there should be a schedule
  return departureInfo.prediction && departureInfo.prediction.time
    ? departureInfo.prediction.time
    : departureInfo.schedule!.time;
};

const getNextUnCancelledDeparture = (
  departureInfos: DepartureInfo[]
): DepartureInfo | undefined => {
  return find(departureInfos, (di: DepartureInfo) => !di.isCancelled);
};

const displayInfoContainsPrediction = (
  departureInfo: DepartureInfo
): boolean => {
  return !!departureInfo.prediction;
};

const mergeIntoDepartureInfo = (
  schedules: ScheduleWithTimestamp[],
  predictions: PredictionWithTimestamp[]
): DepartureInfo[] => {
  const schedulesByTripId = keyBy(schedules, "trip.id");
  const predictionsByTripId = keyBy(predictions, "trip.id");
  const scheduleTripIds = keys(schedulesByTripId);
  const predictionTripIds = keys(predictionsByTripId);

  // All the schedules with and without predictions
  const scheduleDirectionInfo = map(
    schedulesByTripId,
    (schedule, tripId): DepartureInfo => {
      // prediction could be undefined
      const prediction = predictionsByTripId[tripId];
      return {
        prediction,
        schedule,
        route: schedule.route,
        trip: schedule.trip,
        isCancelled: isCancelled(prediction),
        isSkipped: isSkipped(prediction),
        isDelayed: isDelayed(prediction, schedule),
        routeMode: toRouteMode(schedule.route)
      };
    }
  );

  const unmatchedPredictionTripIds = difference(
    predictionTripIds,
    scheduleTripIds
  );
  // All the predictions with no schedules
  const predictionDirectionInfo = map(
    unmatchedPredictionTripIds,
    (tripId: string): DepartureInfo => {
      const prediction = predictionsByTripId[tripId];
      return {
        prediction,
        route: prediction.route,
        trip: prediction.trip,
        isCancelled: isCancelled(prediction),
        isSkipped: isSkipped(prediction),
        routeMode: toRouteMode(prediction.route)
      } as DepartureInfo;
    }
  );

  const departureInfos = concat(scheduleDirectionInfo, predictionDirectionInfo);
  const sortedDepartureInfo = sortBy(departureInfos, di => {
    // prioritize sorting by predictions
    if (di.prediction && di.prediction.time) {
      return di.prediction.time;
    }
    return di.schedule?.time;
  });

  return sortedDepartureInfo;
};

const isAtDestination = (
  stopName: string,
  route: Route,
  directionId: DirectionId
): boolean => {
  const destination = route.direction_destinations[directionId];
  if (destination) {
    return destination.includes(stopName) && !destination.includes(" or ");
  }
  return false;
};

const isCommuterRail = (departureInfo: DepartureInfo): boolean => {
  return departureInfo?.routeMode === "commuter_rail";
};

const DefaultWrapper: React.FunctionComponent<unknown> = ({ children }) => (
  <li>{children}</li>
);
/**
 * From a list of `DepartureInfo[]`, generate a list of `<DisplayTime />`.
 * Each `<DisplayTime />` will recieve the `isCR` and `targetDate` props.
 *
 * This function can optionally
 * - wrap each `<DisplayTime />` in a specified wrapper element, `WrapperEl`
 * - truncate the list to a desired `listLength` value
 */
const departuresListFromInfos = (
  departureInfos: DepartureInfo[],
  isCR: boolean,
  targetDate?: Date,
  listLength?: number,
  omitCancelledAndSkipped = false,
  WrapperEl: typeof DefaultWrapper = DefaultWrapper
): React.ReactElement[] => {
  // optional cutoff time, before which we won't show schedules.
  // just used with subway for now.
  const predictionTimeCutoff = chain(departureInfos)
    .filter(d => d.routeMode === SUBWAY)
    .maxBy("prediction.time")
    .value()?.prediction!.time;

  return chain(departureInfos)
    .reject(
      departure =>
        omitCancelledAndSkipped &&
        (!!departure.isCancelled || !!departure.isSkipped)
    )
    .omitBy(
      ({ prediction, schedule }) =>
        // omit schedule-only departures that are before latest prediction time
        predictionTimeCutoff &&
        !prediction &&
        schedule &&
        schedule.time <= predictionTimeCutoff
    )
    .map(d => (
      <WrapperEl key={getInfoKey(d)}>
        <DisplayTime departure={d} isCR={isCR} targetDate={targetDate} />
      </WrapperEl>
    ))
    .slice(0, listLength)
    .value();
};

const departureInfoInRoutePatterns = (
  departureInfo: DepartureInfo,
  routePatterns: RoutePatternWithPolyline[]
): boolean => {
  const isMatch = !!routePatterns.find(
    rp => rp.id === departureInfo.trip.route_pattern_id
  );
  // This second case is needed for the edge case where the route pattern is
  // canonical but the schedules/predictions are not serving canonical route
  // patterns (e.g. due to shuttles or other disruptions)
  const isSimilar = !!routePatterns.find(
    rp =>
      rp.headsign === departureInfo.trip.headsign &&
      rp.route_id === departureInfo.route.id
  );
  return isMatch || isSimilar;
};

export {
  mergeIntoDepartureInfo,
  departureInfoToTime,
  displayInfoContainsPrediction,
  getNextUnCancelledDeparture,
  isAtDestination,
  isCommuterRail,
  departuresListFromInfos,
  departureInfoInRoutePatterns
};
