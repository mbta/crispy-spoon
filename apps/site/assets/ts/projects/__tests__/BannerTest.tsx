import React from "react";
import renderer from "react-test-renderer";
import { mount } from "enzyme";
import { createReactRoot } from "../../app/helpers/testUtils";
import { Teaser as Project } from "../../__cms";
import Banner, { cmsRouteToClass } from "../components/Banner";

const bannerTeaser: Project = {
  title: "Better Bus Project",
  text:
    "Too many of our bus routes still fail to live up to our own standards. Through the Better Bus Project, we are changing that.",
  status: null,
  routes: [
    {
      mode: "bus",
      id: "35",
      group: "route"
    },
    {
      mode: "bus",
      id: "36",
      group: "route"
    },
    {
      mode: "bus",
      id: "37",
      group: "route"
    },
    {
      mode: "bus",
      id: "52",
      group: "route"
    },
    {
      mode: "bus",
      id: "64",
      group: "route"
    },
    {
      mode: "bus",
      id: "70",
      group: "route"
    },
    {
      mode: "bus",
      id: "70A",
      group: "route"
    },
    {
      mode: "bus",
      id: "72",
      group: "route"
    },
    {
      mode: "bus",
      id: "74",
      group: "route"
    },
    {
      mode: "bus",
      id: "75",
      group: "route"
    },
    {
      mode: "bus",
      id: "201",
      group: "route"
    },
    {
      mode: "bus",
      id: "202",
      group: "route"
    },
    {
      mode: "bus",
      id: "220",
      group: "route"
    },
    {
      mode: "bus",
      id: "222",
      group: "route"
    },
    {
      mode: "bus",
      id: "225",
      group: "route"
    }
  ],
  path: "/projects/better-bus-project",
  image: {
    url:
      "https://live-mbta.pantheonsite.io/sites/default/files/styles/teaser/public/projects/betterbus/betterbus-banner-large.png?itok=qWyWNEZq",
    alt: "Better Bus Project: Making transit better together"
  },
  id: 3445,
  date: "2019-07-30"
};

it("renders", () => {
  createReactRoot();
  const tree = renderer
    .create(
      <Banner banner={bannerTeaser} placeholderImageUrl="path-to-some-image" />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("returns null if no banner project", () => {
  const wrapper = mount(
    <Banner banner={null} placeholderImageUrl={placeholderImageUrl} />
  );
  expect(wrapper.html()).toEqual(null);
});

const placeholderImageUrl = "path-to-some-image";
it("renders with unknown background color if no routes are listed", () => {
  const wrapper = mount(
    <Banner
      banner={{ ...bannerTeaser, routes: [] }}
      placeholderImageUrl={placeholderImageUrl}
    />
  );
  const actualSrc = wrapper.find(".c-banner__content").first();
  expect(actualSrc.html()).toContain("u-bg--unknown");
});

it("renders placeholder image if there is no image", () => {
  const wrapper = mount(
    <Banner
      banner={{ ...bannerTeaser, image: null }}
      placeholderImageUrl={placeholderImageUrl}
    />
  );
  const actualSrc = wrapper.find(".c-banner__image");
  expect(actualSrc.first().html()).toContain(placeholderImageUrl);
  expect(actualSrc.last().html()).toContain(placeholderImageUrl);
});

describe("cmsRouteToClass", () => {
  const route = { mode: "", id: "", group: "" };
  it("handles routes", () => {
    expect(cmsRouteToClass({ ...route, id: "Green" })).toEqual("green-line");
    expect(cmsRouteToClass({ ...route, id: "silver_line" })).toEqual(
      "silver-line"
    );
    expect(cmsRouteToClass({ ...route, id: "mattapan" })).toEqual("mattapan");
    expect(cmsRouteToClass({ ...route, group: "custom", mode: null })).toEqual(
      "unknown"
    );
    expect(cmsRouteToClass({ ...route, group: "custom", mode: "bus" })).toEqual(
      "bus"
    );
    expect(
      cmsRouteToClass({ ...route, group: "mode", mode: "bus", id: "some_mode" })
    ).toEqual("some-mode");
    expect(
      cmsRouteToClass({ ...route, group: "line", mode: "bus", id: "Orange" })
    ).toEqual("orange-line");
    expect(cmsRouteToClass({ ...route, mode: "commuter_rail" })).toEqual(
      "commuter-rail"
    );
  });
});
