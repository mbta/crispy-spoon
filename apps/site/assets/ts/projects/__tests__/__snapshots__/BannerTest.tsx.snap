// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
Array [
  <h2
    className="c-projects-header__subheader"
  >
    Featured Projects
  </h2>,
  <a
    className="c-banner c-banner--responsive c-banner--lg-9 c-banner--no-margin-top c-banner--default"
    href="/projects/better-bus-project"
  >
    <div
      className="hidden-sm-up"
    >
      <div
        className="row"
      >
        <div
          className="c-banner__image c-banner__image--responsive c-banner__image--default"
          style={
            Object {
              "backgroundImage": "url(https://live-mbta.pantheonsite.io/sites/default/files/styles/teaser/public/projects/betterbus/betterbus-banner-large.png?itok=qWyWNEZq)",
            }
          }
        >
          <div
            className="sr-only"
          >
            Better Bus Project: Making transit better together
          </div>
        </div>
        <div
          className="c-banner__content c-banner__content--responsive-side-by-side c-banner__content--left u-bg--unknown"
        >
          <div
            className="c-banner__top"
          >
            <h2
              className="h2 c-banner__title c-banner__title--default"
            >
              Better Bus Project
            </h2>
          </div>
          <div
            className="c-banner__bottom"
          >
            <span
              className="c-banner__date"
            >
              Updated 
              July 30, 2019
            </span>
          </div>
        </div>
      </div>
    </div>
    <div
      className="hidden-xs-down hidden-lg-up"
    >
      <div
        className="row"
      >
        <div
          className="c-banner__image c-banner__image--responsive-side-by-side c-banner--responsive-no-margin c-banner__image--default"
        >
          <div
            className="c-banner__content c-banner__content--responsive-side-by-side c-banner__content--left u-bg--unknown"
          >
            <div
              className="c-banner__top"
            >
              <h2
                className="h2 c-banner__title c-banner__title--default"
              >
                Better Bus Project
              </h2>
            </div>
            <div
              className="c-banner__bottom"
            >
              <span
                className="c-banner__date"
              >
                Updated 
                July 30, 2019
              </span>
            </div>
          </div>
          <img
            alt="Better Bus Project: Making transit better together"
            className="c-banner__image--by-side"
            src="https://live-mbta.pantheonsite.io/sites/default/files/styles/teaser/public/projects/betterbus/betterbus-banner-large.png?itok=qWyWNEZq"
          />
          <div
            className="sr-only"
          >
            Better Bus Project: Making transit better together
          </div>
        </div>
      </div>
    </div>
    <div
      className="hidden-md-down"
    >
      <div
        className="c-banner__image c-banner__image--responsive-side-by-side c-banner--responsive-no-margin c-banner__image--default"
      >
        <div
          className="c-banner__content c-banner__content--responsive-side-by-side c-banner__content--left u-bg--unknown"
        >
          <div
            className="c-banner__top"
          >
            <h2
              className="h2 c-banner__title c-banner__title--default"
            >
              Better Bus Project
            </h2>
          </div>
          <div
            className="c-banner__bottom"
          >
            <span
              className="c-banner__date"
            >
              Updated 
              July 30, 2019
            </span>
          </div>
        </div>
        <img
          alt="Better Bus Project: Making transit better together"
          className="c-banner__image--by-side"
          src="https://live-mbta.pantheonsite.io/sites/default/files/styles/teaser/public/projects/betterbus/betterbus-banner-large.png?itok=qWyWNEZq"
        />
        <div
          className="sr-only"
        >
          Better Bus Project: Making transit better together
        </div>
      </div>
    </div>
  </a>,
]
`;
